pkg/guestagent/events/containerd_linux.go

```
@@ -0,0 +1,369 @@
+// SPDX-FileCopyrightText: Copyright The Lima Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package events
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/api/events"
+	"github.com/containerd/containerd/errdefs"
+	containerdNamespace "github.com/containerd/containerd/namespaces"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/lima-vm/lima/v2/pkg/guestagent/api"
+)
+
+const (
+	stateKey             = "nerdctl/state-dir"
+	portsKey             = "nerdctl/ports"
+	namespaceKey         = "nerdctl/namespace"
+	defaultSocketTimeout = 5 * time.Second
+)
+
+type ContainerdEventMonitor struct {
+	socketPaths            []string
+	runningContainersMutex sync.Mutex
+	runningContainers      map[string][]*api.IPPort
+}
+
+func NewContainerdEventMonitor(socketPaths []string) *ContainerdEventMonitor {
+	return &ContainerdEventMonitor{
+		socketPaths:       socketPaths,
+		runningContainers: make(map[string][]*api.IPPort),
+	}
+}
+
+func (c *ContainerdEventMonitor) MonitorPorts(ctx context.Context, ch chan *api.Event, errCh chan error) {
+	var wg sync.WaitGroup
+	for _, socket := range c.socketPaths {
+		wg.Add(1)
+		go func(socket string) {
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					errCh <- ctx.Err()
```

I'm not sure we need to report an error here; this should only happen when the process has been asked to shut down.

If we want to log the information, then it should include information that allows us to identify the goroutine, e.g. by including the socket name.